### DISCUSSION TITLE
DataCollection default to empty array value

### DISCUSSION THREAD

**Discussion by mason-rogers** [spatie/laravel-data#336] [Q&A]
Hi, I am migrating from Spatie\DataTransferObject to LaravelData, I'm having some trouble migrating one of our DTO's.

DataTransferObject had the ability to cast an array type's inner values to a class, using eg. `#[CastWith(ArrayCaster::class, itemType: ConditionData::class)]`, but the actual type remained as `array`. eg. `public array $conditions = [];`

But with DataCollection, I can not set the default value to `[]` as it's a mismatched type. I am wondering what the equivalent for that is in LaravelData - because I also can not simply `= ConditionData::collection()`, as it's an expression

# Snippet of DTO after migration

```php
class ConditionData extends Data
{
    public bool $group = false;

    public string $operand = Ruleset::OPERAND_AND;

    public ?string $type;

    public ?string $attribute;

    public ?string $operator;

    public null|string|bool|array $value;

    /**
     * @var DataCollection<ConditionData>
     */
    #[DataCollectionOf(ConditionData::class)]
    public DataCollection $conditions;

}
```

# Old usage (which worked previously):
```php
$condition = new ConditionData([
    'group' => true,
    'operand' => 'and',
]);

$condition->conditions // []
```

# Ideal New usage (without `conditions` array passed, throws error):
```php
$condition = ConditionData::from([
    'group' => true,
    'operand' => 'and',
]);

$condition->conditions -> // Typed property ConditionData::$conditions must not be accessed before initialization
```

# New usage without `conditions` array passed:
```php
$condition = ConditionData::from([
    'group' => true,
    'operand' => 'and',
    'conditions' => [],
]);

$condition->conditions -> // []
```

**Comment by rubenvanassche** [comment]
Hmm such thing is not possible at the moment, one of my plans is to get completely rid of the DataCollection type in the future and just use arrays but that's something which will take a lot of time.

For now, You could use `prepareForPipeline` to add the array as a default property if it not has been set.

Or create a custom pipe which checks each DataProperty if it is a data collection, if not set, an empty array can be set.

**Comment by Defrothew** [comment]
I'm also looking for similar behavior for this. I ended up doing like this.

```
public function __construct(
    #[DataCollectionOf(ConditionData::class)]
    public ?DataCollection $conditions = null
    )
{
  $this->conditions = $conditions ?: new DataCollection(ConditionData::class, []);
}
```

Is this approach valid and does not have any side effects with other features such as using in resource and request validation?

**Comment by igorsgm** [comment]
I'm also wondering what's the best way to achieve it. In nested data classes, I feel like @Defrothew is an easier approach. But I'm not sure about the side effects 🤔
